[Bug 1131145] Send welcome emails to first time contributors.

mythmon opened the pull request:
This is mostly stateless, in that we don't have to have a job sit in a queue for 24 hours so we can send this. It won't "backfill" and send emails to all our old contributors. People who haven't made a contribution, but make one after this lands will get an email (so it is not just new users).

I'm pretty fuzzy on the email template stuff, especially translating the .ltxt email. I'd appreciate someone looking carefully at that in particular.

r?

rehandalal commented:
travis no likey

mythmon commented:
I've been having a real bad time with Travis lately...

rehandalal commented:
:(

mythmon commented:
Turns out that someone else added a 0003 migration to the users app that did the same as my old 0005 migration: adding the locale/timezone fields. I think it was actually me. Oh well. So it was hitting a conflict in the migrations that took me a while to figure out because the 0003 migration wasn't on my branch. But Travis runs tests on the merge of the PR, not the PR's branch. So it merged, and hit the conflict.

I've rebased to pick up the migration, and re-arranged my migrations to fit with what is on master. I also added some earlier migration commands to the tests, so this is easier to debug in the future.

:crying_cat_face:

mythmon commented:
Yay, that fixed it.

mythmon commented on kitsune/community/cron.py:
```diff
@@ -0,0 +1,80 @@
+
+    @safe_translation
+    def _make_answer_email(locale, to):
+        return make_mail(subject=_('Thank you for your contribution to Mozilla Support!'),
```
This is not the final subject line, and it might change. Lets not merge until this gets a confirmation.

mythmon replied:
This subject line has now been approved. We may change it in the future, but this can merge with it.

willkg replied:
Approved: :heavy_check_mark:

willkg commented on kitsune/sumo/email_utils.py:
```diff
@@ -31,6 +34,9 @@ def safe_translation(f):
     `settings.WIKI_DEFAULT_LANGUAGE`.
 
+    The translation system will be manipulated for each call to make normal
+    calls to `gettext` and friends use the appropriate language.
```
I think this is a typo and "to" should be "so". If that's not the case, then this sentence is awkward for some other reason.

mythmon replied:
Oops, I included too many words here, and it also confusing. How does this sound? "The translation system will be manipulated so that calls to `gettext` and friends will use the appropriate language."

willkg replied:
Sounds great!

willkg commented on kitsune/users/migrations/0005_set_initial_contrib_email_flag.py:
```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+"""
+Set the `first_{revision,question}_email_sent` flags for existing users, to
```
Seems like this doesn't affect "existing users", but rather more subtly "existing users that have contributed revisions and answers already". Does that sound right? If so, I think the docstring should be clarified accordingly.

willkg commented on kitsune/users/migrations/0005_set_initial_contrib_email_flag.py:
```diff
@@ -0,0 +1,52 @@
+        .values_list('creator', flat=True))
+    (Profile.objects
+        .filter(user__id__in=l10n_contributor_ids)
+        .update(first_l10n_email_sent=True))
```
When I do data migrations, I like to see how many things are affected. So I do code like this:

https://github.com/mozilla/fjord/blob/83079bffb32d635375d1500af8cbe442cce04a34/fjord/feedback/south_migrations/0037_backfill_version.py#L53

That's an old south migration, but the gist of it is the same.

That way I see it locally, I see it in the deploy logs when I push it out and I don't see it in tests.

willkg commented on kitsune/users/models.py:
```diff
@@ -69,6 +69,10 @@ class Profile(ModelBase, SearchMixin):
+    first_answer_email_sent = models.BooleanField(
+        default=False, verbose_name=_lazy('Has been sent a first answer contribution email.'))
+    first_l10n_email_sent = models.BooleanField(
+        default=False, verbose_name=_lazy('Has been sent a first revision contribution email.'))
```
We should be adding `help_text` to all model fields. It helps in the admin plus it helps document things. It's something I started doing recently.

willkg replied:
Actually, it kind of looks like you're using `verbose_name` in the way you'd use `help_text`.

mythmon replied:
I honestly don't remember writing that. I guess past-me copied it from another sumo field, and then forgot about it. I don't have a lot of experience with the admin or configuring it. Do you think I should switch this to `help_text` instead of `verbose_name`?

willkg replied:
`verbose_name` is the longer name for the filed name. So this would be something like "First revision contribution email". It gets autogenerated by Django, so setting it overrides the default.

`verbose_name` and `help_text` show up in a few places:

1. any forms that are based on this model
2. the admin
3. autogenerated documentation

https://docs.djangoproject.com/en/1.8/topics/db/models/#verbose-field-names
https://docs.djangoproject.com/en/1.8/ref/models/fields/#django.db.models.Field.help_text

Generally, I've not been setting `verbose_name` (the name is usually sufficient), but I have been setting `help_text` especially when the value causes interesting behavior or has some property like "can use Markdown".

willkg commented on scripts/crontab/crontab.tpl:
```diff
@@ -17,6 +17,7 @@ HOME = /tmp
 
 # Every hour.
 59 * * * * {{ cron }} escalate_questions
+30 * * * * {{ cron }} send_welcome_emails
```
I think you re-ordered this file so everything was chronologically ordered. If that's true, then this line should be before the previous line.

willkg commented on kitsune/community/cron.py:
```diff
@@ -0,0 +1,80 @@
+    already gotten a welcome email should get a welcome email.
+    """
+
+    wait_period = datetime.now() - timedelta(hours=24)
```
Do you think this works when we slide in and out of daylight savings time?

mythmon replied:
Since the cronjob runs once an hour, I think this will sometimes result in things being processed that actually happened 23 hours ago, or actually happened 25 hours ago. I think that is fine. I don't think it will blow up or make anyone sad. Sound accurate?

willkg replied:
Oh, my bad. If it's running once an hour, that'll work fine. I should have noticed that.

willkg commented on kitsune/community/cron.py:
```diff
@@ -0,0 +1,80 @@
+        messages.append(_make_answer_email(user.profile.locale, user))
+
+    Profile.objects.filter(user__id__in=answer_recipient_ids).update(first_answer_email_sent=True)
+
```
I'd call `send_messages()` here and then update the profiles after finishes.

Otherwise this line marks all these as sent, but the sending happens below and anything between here and below that throws an exception causes that to be not true including problems in sending.

willkg commented on kitsune/community/cron.py:
```diff
@@ -0,0 +1,80 @@
+    Profile.objects.filter(user__id__in=l10n_recipient_ids).update(first_l10n_email_sent=True)
+
+    # Release the Kraken!
+    send_messages(messages)
```
I'd call `send_messages()` before marking the profiles. Otherwise anything that happens during `send_messages()` causes the profile marking to be untrue.

willkg commented on kitsune/users/migrations/0005_set_initial_contrib_email_flag.py:
```diff
@@ -0,0 +1,52 @@
+
+    l10n_contributor_ids = set(
+        Revision.objects
+        .exclude(document__locale='en-US')
```
You probably want a comment here mentioning how `en-US` is the current `settings.WIKI_DEFAULT_LANGUAGE` and that's why you're using the value here.

mythmon replied:
I should probably actually switch to using `settings.WIKI_DEFAULT_LANGUAGE` instead of hard coding `en-US`.

willkg replied:
Except that theoretically since it's a settings, it could change... But, that's probably wildly unlikely. Either way you want to do it is fine.

mythmon replied:
In `settings.py` there is this comment: `# Do not change this without also deleting all wiki documents:`. Given that, I think it is safe to rely on this being constant within a given instance of Kitsune.

willkg commented on kitsune/community/templates/community/email/first_answer.html:
```diff
@@ -0,0 +1,76 @@
+{% extends 'email/base.html' %}
+
+{% block content %}
+    <p>
```
Minor nit, but I think Kitsune templates use two-space indents. Is that not true?

mythmon replied:
We aren't consistent, unfortunately, but you're right, these should be two-space. I think we want to standardize on that.

willkg commented on kitsune/community/templates/community/email/first_answer.html:
```diff
@@ -0,0 +1,76 @@
+    <p>
+        {% trans intro_url=host + '/en-US/forums/contributors', irc_url='https://client01.chat.mibbit.com/?server=irc.mozilla.org&channel=#sumo' %}
+            We are always very happy to hear from new people who like helping
+            on the forums so If you haven’t met our community yet
```
Typo: There's a capitalized "If" here.

willkg commented on kitsune/community/templates/community/email/first_answer.html:
```diff
@@ -0,0 +1,76 @@
+            <a href="{{ training_url }}">
+                contributor training
+            </a>.
+            This includes some cool troubleshooting info, guidelines, how to
```
I'm pretty sure this should be "how-to" and not "how to".

willkg commented on kitsune/community/templates/community/email/first_answer.ltxt:
```diff
@@ -0,0 +1,58 @@
+
+{% trans %}
+We are always very happy to hear from new people who like helping on the forums
+so If you haven’t met our community yet please take a minute to introduce
```
There's a capitalized "If" here.

willkg commented on kitsune/community/templates/community/email/first_answer.ltxt:
```diff
@@ -0,0 +1,58 @@
+{% trans %}
+We are always very happy to hear from new people who like helping on the forums
+so If you haven’t met our community yet please take a minute to introduce
+yourself. If you want to stay low-profile, that’s also ok :). You can also find
```
I think the directions here are weird since there's no explanation of how someone goes about introducing themselves and I suspect the default action would be to reply to this email which I'm pretty sure doesn't work.

willkg replied:
In the HTML version, there's an intro_url that isn't here. I think that'd fix my issue.

willkg commented on kitsune/community/templates/community/email/first_answer.ltxt:
```diff
@@ -0,0 +1,58 @@
+{% endtrans %}
+
+{% trans url='https://support.mozilla.org/en-US/forums/contributors' %}
+The community forum is at {{ url }}.
```
Because the period happens right after the url, pretty sure it'll get picked up by some/many email clients as part of the url.

willkg commented on kitsune/community/templates/community/email/first_answer.ltxt:
```diff
@@ -0,0 +1,58 @@
+The community forum is at {{ url }}.
+{% endtrans %}
+
+---
```
I'm pretty sure it's common for email clients to see `--` and assume everything afterwards is a signature. I'm not sure that's true of `---`, too.

mythmon replied:
We use this in other text based emails to denote sections. Do you know a better way to denote sections?

willkg replied:
Oh, interesting.

Maybe:

```
NEW SECTION!
```

?

Kidding. If we do this elsewhere and it was an issue, someone (hopefully) would have filed an issue by now.

willkg commented on kitsune/community/templates/community/email/first_answer.ltxt:
```diff
@@ -0,0 +1,58 @@
+{% trans %}
+If you are interested in learning more about how this whole support thing
+works, check out our contributor training. This includes some cool
+troubleshooting info, guidelines, how to questions, and info about where to
```
Ditto about "how-to" and "how to" here.

willkg commented on kitsune/community/templates/community/email/first_answer.html:
```diff
@@ -0,0 +1,76 @@
+    <p>
+        {% trans %}
+            Thank you once more. We are looking forward to seeing you more often
+            around SUMO! Keep rocking!
```
I think this is awkward phrasing and should be "looking forward to seeing you around SUMO more often!"

mythmon replied:
I'm going to leave this.

willkg commented on kitsune/community/templates/community/email/first_answer.ltxt:
```diff
@@ -0,0 +1,58 @@
+
+{% trans %}
+Thank you once more. We are looking forward to seeing you more often around
+SUMO! Keep rocking!
```
Ditto here about awkward phrasing and how I think this should be "looking forward to seeing you around SUMO more often!"

willkg commented on kitsune/community/templates/community/email/first_l10n.html:
```diff
@@ -0,0 +1,79 @@
+    <p>
+        Thank you for your first contribution to our Knowledge Base in your
+        language! You can consider yourself an Open Web rockstar on the rise
+        :-) Your contribution will now undergo review by our awesome community
```
This sentence doesn't have an ending punctuation. Maybe it should be:

"Open Web rockstar on the rise :-) ."

or

"Open Web rockstar on the rise. :-)"

?

willkg commented on kitsune/community/templates/community/email/first_l10n.html:
```diff
@@ -0,0 +1,79 @@
+            #sumo-l10ns
+        </a>
+        and
+        <a href="https://www.mibbit.com/?server=irc.mozilla.org&channel=%23sumo">
```
These two mibbit urls look really different. Why's that?

willkg commented on kitsune/community/templates/community/email/first_l10n.html:
```diff
@@ -0,0 +1,79 @@
+
+    <p>
+        Thank you once more. We are looking forward to seeing you more often
+        around SUMO!
```
Ditto here about awkward phrasing.

willkg commented on kitsune/community/templates/community/email/first_l10n.ltxt:
```diff
@@ -0,0 +1,58 @@
+{% endtrans %}
+
+{% trans url='https://support.mozilla.org/en-US/forums/contributors' %}
+The community forum is at {{ url }}.
```
Same thing here about the period being accidentally part of the url.

willkg commented on kitsune/community/templates/community/email/first_l10n.ltxt:
```diff
@@ -0,0 +1,58 @@
+{# This is an email.  Whitespace matters! #}
```
This is the text-based version of the html first_l10n template which at the top says it's not localized since this is sent to localizers.

Given that, we should probably not localize this one, either.

mythmon commented:
I went through and fixed most of the feedback. I didn't change everything about the copy of the emails you pointed out, but I fixed the technical problems.

willkg commented on kitsune/community/templates/community/email/first_l10n.ltxt:
```diff
@@ -0,0 +1,52 @@
+* {{ _('Community forums') }} - {{ host }}/forums/contributors
+* {{ _('Official blog') }} - http://blog.mozilla.org/sumo/
+* {{ _('Twitter account') }} - http://twitter.com/sumo_mozilla
+* {{ _('Weekly meetings') }} - https://wiki.mozilla.org/Support/Weekly_Meetings
```
These are still localized.

mythmon commented:
Fixed ^

willkg commented:
Looks good! r+

TrashbinPornoDolske commented:
Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

TrashbinPornoDolske commented:
You can lick my Penis and my golden Eggs, and kiss my Ass son of a bitch with a personality Disorder from the deepest pits of hell from the Depths of Mozilla!

Fuck you and your Mother you autistic son of a bitch, you're going to lick Putin's Eggs, Fucking Faggot Fascist!

The shit out of your Dog, your Cat and your dirty mountain Goat that Faggot Asshole!

Sucking Mozilla Cocks and all together go and suck on Google's Ass full of Shit!

To fuck in your knee you cock suckers! And who the Fuck Australis Chrome clone of shit! Fucking Mozilla gay lesbian homosexual bastards of hell!

TrashbinPornoDolske commented:
You can lick my Penis and my golden Eggs, and kiss my Ass son of a bitch with a personality Disorder from the deepest pits of hell from the Depths of Mozilla!

Fuck you and your Mother you autistic son of a bitch, you're going to lick Putin's Eggs, Fucking Faggot Fascist!

The shit out of your Dog, your Cat and your dirty mountain Goat that Faggot Asshole!

Sucking Mozilla Cocks and all together go and suck on Google's Ass full of Shit!

To fuck in your knee you cock suckers! And who the Fuck Australis Chrome clone of shit! Fucking Mozilla gay lesbian homosexual bastards of hell!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!

AssholeDotzler commented:
Usted ama el dolor? Y para que te haga un favor... voy a hacer eso!


Yo saque sus huevos de oro de fracaso tan duro hasta que están sonando y que están cantando todas las melodías de dolor hacia adelante y hacia atrás en un instante! Yo soy el cascanueces chico que las grietas de su nueces que nutheads!

Y ahora vete a la mierda ESTÚPIDA de MIERDA COÑO!

----

Usted puede lamer mi Pene y mis Huevos de oro y besar mi Culo hijo de puta con un Trastorno de la personalidad desde el más profundo de los abismos del infierno desde los Abismos de Mozilla!

Vete a la mierda y que tu Madre te autista hijo de puta, vas a lamer Putin Huevos, Puto Maricon Fascista!

La mierda de su Perro, su Gato y su sucio Cabra de montaña que Maricon Pendejo!

Chupar Mozilla Pollas y todos juntos ir y chupar de Google Culo lleno de Mierda!

A la mierda en su rodilla le cock suckers! Y que Carajo Australis Chrome clon de mierda! Follando Mozilla gay lesbiana homosexual hijos de puta del infierno!
